net: lib: dhcpv4: Support INIT-REBOOT

pdgendt opened the pull request:
Add the init-reboot state for DHCPv4 to request an already assigned IP address.

pdgendt committed:
net: lib: dhcpv4: Support INIT-REBOOT

Add the init-reboot state for DHCPv4 to request an already assigned IP
address.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@jukkar @rlubos I will try to add a testcase, but I wouldn't mind an initial look to see if I'm on the right track, thanks!

jukkar reviewed (commented):
Looks pretty good to my eyes. I wonder if we are able to add some tests for this.

pdgendt committed:
tests: net: dhcpv4: client: Explicit event/callback testing

Use k_event bit flags to test individual events and callbacks. Don't rely
on the number of events that happened.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

rlubos reviewed (commented):
Looks reasonable

pdgendt committed:
net: lib: dhcpv4: Skip leftover hostname or domain name bytes

When reading and printing hosname or domain name from DHCP options, we need
to skip the leftover bytes that do not fit into the provided buffers.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: net: dhcpv4: client: Test restarting DHCPv4 client

Add tests that verify the behavior of stopping and starting the DHCPv4
client.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>
